# Karyotyping
This is the updated karyotyping pipeline for Sheltzer Lab's publications. It uses CNVKit for inferring copy numbers and a modified plotting function from the same tool. 

# Data files

# Running the script

To run the script, you can directly run the following command once you have downloaded the data files.
```
    nextflow run https://github.com/sheltzer-lab/Karyotyping.git
```

# Citations

1. Talevich, E., Shain, A.H., Botton, T., & Bastian, B.C. (2014). CNVkit: Genome-wide copy number detection and visualization from targeted sequencing. PLOS Computational Biology 12(4):e1004873
2. i) Olshen, A.B., Bengtsson, H., Neuvial, P., Spellman, P.T., Olshen, R.A., & Seshan, V.E. (2011). Parent-specific copy number in paired tumor-normal studies using circular binary segmentation. Bioinformatics 27(15):2038–46.
  ii) Venkatraman, E.S., & Olshen, A.B. (2007). A faster circular binary segmentation algorithm for the analysis of array CGH data. Bioinformatics 23(6):657–63
